Feature/some kind of card

henrikhermansen opened the pull request:
I wanted to reopen #262 , but because I replaced my previous commits with new ones, I wasn't allowed to do that.

I've taken comments, opinions and issues raised in the previous PR into consideration when I reworked the styling and components for the new cards. With this, I hope we can come a lot closer to being able to merge this.

I've also taken the liberty of naming them. While these are mere suggestions, I do feel they make sense, as they reflect what's significant about each card.
Speaking of names, I have controversially named the base class and component Card Base. My thought behind this is that it's not really a card - it's a base. This also distinguishes it more clearly from the actual card components. Thoughts?

Closes #260

selbekk commented:
Can we get a screenshot of how this will look? :)

henrikhermansen commented:
![image](https://user-images.githubusercontent.com/435037/48347652-22918f80-e67f-11e8-8113-4a6fb72a0a6c.png)
![image](https://user-images.githubusercontent.com/435037/48347656-26bdad00-e67f-11e8-982c-e8fd107198e4.png)
![image](https://user-images.githubusercontent.com/435037/48347659-29200700-e67f-11e8-850b-2dbe59c3ed83.png)
![image](https://user-images.githubusercontent.com/435037/48347661-2c1af780-e67f-11e8-871a-10e12f85fd89.png)
![image](https://user-images.githubusercontent.com/435037/48347678-3ccb6d80-e67f-11e8-82e4-075bfd10a3be.png)
![image](https://user-images.githubusercontent.com/435037/48347682-3fc65e00-e67f-11e8-8e9f-c0b1ff6edfe8.png)
![image](https://user-images.githubusercontent.com/435037/48347687-4228b800-e67f-11e8-9bd2-8e564637ef79.png)

ivarni commented on packages/ffe-cards-react/src/IconCard/IconCard.md:
```diff
@@ -0,0 +1,49 @@
+Ikonkort er kort med en gitt maksbredde og et venstrestilt ikon.
+Den tilbyr alle underkomponenter (CardName, Title, Subtext og Text) og plasserer dem i
+gitt rekkefølge nedover i kortet, med teksten venstrestilt, til høyere for ikonet.
```
*høyre

henrikhermansen replied:
Sant, fixed

ivarni commented on packages/ffe-cards-react/src/IconCard/IconCard.md:
````diff
@@ -0,0 +1,49 @@
+```jsx
+const { GrafOppIkon } = require('../../../ffe-icons-react');
+
+<React.Fragment>
````
Tror ikke du trenger å wrappe `IconCard` i `Fragment`?

henrikhermansen replied:
Nei, riktig. Jeg hadde to på et tidspunkt, og glemte å fjerne fragment da jeg fjernet det andre kortet. Fixet nå.

ivarni commented on packages/ffe-cards-react/src/ImageCard/ImageCard.md:
```diff
@@ -0,0 +1,43 @@
+        }
+    >
+        {({ Title }) => (
+            <React.Fragment>
```
Her trenger du heller ikke `Fragment`

henrikhermansen replied:
Good catch! Jeg hadde flere subcomponents på et tidspunkt. Glemte å fjerne Fragment da jeg gikk ned til bare det ene. Fixed.

selbekk commented on packages/ffe-cards-react/src/IconCard/IconCard.md:
````diff
@@ -0,0 +1,49 @@
+        )}
+    </IconCard>
+</React.Fragment>;
+```
````
I guess you could drop Fragment around `<IconCard />`

selbekk replied:
Go through all the examples to ensure they're all wrapper-free if they don't require it. Makes for better reading when you look at the example code :)

henrikhermansen replied:
Yeah, I saw there was one more in TextCard.md. Fixed locally.

selbekk commented on packages/ffe-cards-react/src/ImageCard/ImageCard.js:
```diff
@@ -0,0 +1,33 @@
+    className: string,
+    /** A rendered image */
+    image: node.isRequired,
+    children: oneOfType([func, node]),
```
```suggestion
    /** Function that's passed available sub-components as arguments, or regular children */
    children: oneOfType([func, node]),
```

henrikhermansen replied:
Thanks, fixed in all of the three cards

selbekk commented on packages/ffe-cards-react/src/ImageCard/ImageCard.md:
```diff
@@ -0,0 +1,43 @@
+<React.Fragment>
+    <ImageCard
+        image={
+            <img src="https://www.sparebank1.no/content/dam/SB1/foto/profilbilder-liggende/ung-i-sofa.jpg.thumb.1280.1280.jpg" />
```
```suggestion
            <img src="https://www.sparebank1.no/content/dam/SB1/foto/profilbilder-liggende/ung-i-sofa.jpg.thumb.1280.1280.jpg" alt="" />
```

henrikhermansen replied:
Fixed

selbekk commented on packages/ffe-cards-react/src/ImageCard/ImageCard.md:
```diff
@@ -0,0 +1,43 @@
+        image={
+            <img
+                src="https://www.sparebank1.no/content/dam/SB1/foto/profilbilder-liggende/ung-i-sofa.jpg.thumb.1280.1280.jpg"
+                style={{ position: 'relative', bottom: '50px' }}
```
What's this for?

henrikhermansen replied:
Repositioning the image within the boundary, as the parent has restricted width and height.
It's just to show how easy it is to move it around :)

selbekk commented on packages/ffe-cards-react/src/ImageCard/ImageCard.md:
```diff
@@ -0,0 +1,43 @@
+    <ImageCard
+        image={
+            <img
+                src="https://www.sparebank1.no/content/dam/SB1/foto/profilbilder-liggende/ung-i-sofa.jpg.thumb.1280.1280.jpg"
```
```suggestion
                src="https://www.sparebank1.no/content/dam/SB1/foto/profilbilder-liggende/ung-i-sofa.jpg.thumb.1280.1280.jpg"
                alt=""
```

henrikhermansen replied:
Fixed

selbekk commented on packages/ffe-cards-react/src/components/Subtext.js:
```diff
@@ -0,0 +1,16 @@
+
+import ComponentBase from './ComponentBase';
+
+const Subtext = ({ className, ...rest }) => (
```
Is `<SubTitle />` a better name? Subtext is, according to [Wikipedia](https://en.wikipedia.org/wiki/Subtext):

> Subtext or undertone is any content of a creative work which is not announced explicitly by the characters or author, but is implicit or becomes something understood by the observer of the work as the production unfolds.

henrikhermansen replied:
Good question. I took all the sub component names from @adidick's https://github.com/SpareBank1/designsystem/pull/262#issuecomment-392735553

adidick-zz replied:
These names are totally up for improvement so I'm open to any suggestions that make more sense from your perspective.

henrikhermansen replied:
To help reduce misunderstandings, I think it's an advantage if ux and developers use the same component names as much as possible. Whether it be [Subtitle](https://en.wikipedia.org/wiki/Subtitle_(titling)), Subtext or something else.

Any other opinions on which name we should favor?

ivarni replied:
We call them "lead paragraphs" in the regular typography but I suppose that kind of depends on how much text we're going to be putting into them.

selbekk reviewed (dismissed):
A few typos and minor stuff, but I like the approach!

I think the pattern of providing sub-components as props was pretty interesting. It's a new approach, and I like it - but does it bring any benefits over just exporting them as a part of the main API?

henrikhermansen commented:
> does it bring any benefits over just exporting them as a part of the main API?

Good question. While I currently export all sub components through the render prop, I imagine we might make some cards which only allow certain sub components. This way makes it easy to perform such limitations.

On the other hand. If this is never going to be the case, I'm not sure which approach is better. One might argue the API gets a little obfuscated like this, as you won't get any help from the IDE as to which components you have available.

ivarni commented:
I like that it's harder to use the sub components outside of the context they're meant to be used in (cards). The more limited an exposed API can be the better in my opinion, and even if some cards might not have any restrictions on subcomponents this makes it less tempting to use them outside of and actual card-implementation.

selbekk commented:
A different issue - P-tags are not allowed inside BUTTON tags, according to the internet: https://stackoverflow.com/questions/26002381/p-tag-in-button-tag-not-allowed

If the card is a button, we would need to change the default element of each of the sub-components -they need to be spans or something similar.

henrikhermansen commented:
> A different issue - P-tags are not allowed inside BUTTON tags, according to the internet: https://stackoverflow.com/questions/26002381/p-tag-in-button-tag-not-allowed
> 
> If the card is a button, we would need to change the default element of each of the sub-components -they need to be spans or something similar.

Good point. I changed them to be spans as default with display block. Looks exactly the same. I don't know if there are any semantic problems to that approach, but I hope not.

selbekk commented:
I think we should be good. Of course, the client could screw this up themselves, but I think React should issue a warning. 

Speaking of - are there any relevant aria-tags we should add to these?

ivarni commented:
Isn't a `<span>` with display block basically a `<div>`?

selbekk commented:
According to the spec, buttons can only have what's called "phrasing content" inside of it (read more [here](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content)). In other words, to follow the spec, you can't use a `<div />`, because it falls in a different category. It'll work, for sure, but the "correct" way to do it is to use spans.

antidecaf reviewed (commented):
Is a card really a button? This doesn't make sense to me. Buttons only allow "phrasing content" as pointed out above, which means it's basically allowed to contain one sentence of text - not subtitles, paragraphs, etc. Sure we can technically get around the spec using spans in stead of divs and what not, but at that point it feels like the markup is designed to circumvent the semantic representation of the contents.

ivarni commented:
I think in every single usecase I've had for cards so far links have been the obvious choice over buttons since they're anchors for for other pages (detail-pages, buy-our-product-pages etc) so at least from my standpoint it's not really an issues since I can provide an `element` (which 95% of the time for me would be a react-router `Link`).

The example images above also seem like navigation components to me. In what situations are we going to use these as buttons?

selbekk commented:
That's probably a question for @adidick - what's the intended usage of these outside of linking to other content?

henrikhermansen commented:
Good points are being made here! To be honest I can't remember why I defaulted these to buttons. I have the same experience as @ivarni , the cards are pretty much always used with Link or anchor (I can't think of a single example where we used something else).

While they will often be used with element=Link, I wonder what should be the default element? I could put an anchor there, but that calls for a default href too, I suppose. Is it ok to set div as default? Perhaps not have any default element, but require one from the consumer?

ivarni commented:
I think `<div>` is fine, I'm a fan of working defaults whereever possible.

henrikhermansen commented:
I just pushed the recently discussed changes: components are now paragraphs, without display styling, and cards are now divs.

This looks almost the same, but the CardBase now has left-aligned text (which is default for divs, while buttons defaults it to center).
TextCard and ImageCard has explicit styling for centering the text.

adidick-zz commented:
> That's probably a question for @adidick - what's the intended usage of these outside of linking to other content?

That is the only intended use.

Description that lives in the sketch template says:
"Kort er en stor klikkflate, altså en inngang til mer informasjon. Kort har en skygge som alltid skal vises, og blir mer fremtredende på hover og focus state."

selbekk commented on packages/ffe-cards-react/src/CardBase.spec.js:
```diff
@@ -0,0 +1,33 @@
+const children = <div>Hello world</div>;
+
+describe('CardBase', () => {
+    it('should render a button element with correct class and children inside', () => {
```
No longer renders a button - could you please update the text as well?

henrikhermansen replied:
Good catch! For sure, will do that when I squash commits :)
(I'll do the deprecation note for the old cards as well)

selbekk reviewed (dismissed):
Things are looking good!

Could you also please add a note on the deprecated cards that they are, in fact, deprecated?

Commit:
feat(ffe-cards): add styling for new cards

This commit adds styling for three new types of cards and a base class, on
which the other cards are built. It also adds styling for generic
subcomponents to these cards. The card types added here are icon, image
and text.

Commit:
feat(ffe-cards-react): add new card components

This commit adds components for three new cards, as well as a card base
which can be used as a frame on which to build other cards. The commit
also adds some sub components. They are not supposed to be imported by
consumers, but instead are returned to a render function when such is
given to the cards as children. The cards that are added now will all
return all sub components, but we might want to restrict this for certain
cards at some point.

henrikhermansen commented:
Deprecation note added, fixed test description text and squashed commits. All ready to be merged :)
